lib/preopt/src/constant_folding.rs

```diff
@@ -0,0 +1,269 @@
+            (ConstImm::Ieee32(imm0), ConstImm::Ieee32(imm1)) => Some(ConstImm::Ieee32(imm0 - imm1)),
+            (ConstImm::Ieee64(imm0), ConstImm::Ieee64(imm1)) => Some(ConstImm::Ieee64(imm0 - imm1)),
+            _ => unreachable!(),
+        },
+        ir::Opcode::Fmul => match (imm0, imm1) {
+            (ConstImm::Ieee32(imm0), ConstImm::Ieee32(imm1)) => Some(ConstImm::Ieee32(imm0 * imm1)),
+            (ConstImm::Ieee64(imm0), ConstImm::Ieee64(imm1)) => Some(ConstImm::Ieee64(imm0 * imm1)),
+            _ => unreachable!(),
+        },
+        ir::Opcode::Fdiv => match (imm0, imm1) {
+            (ConstImm::Ieee32(imm0), ConstImm::Ieee32(imm1)) => {
+                if imm1 == 0.0 {
+                    return None;
+                }
+                Some(ConstImm::Ieee32(imm0 / imm1))
+            }
+            (ConstImm::Ieee64(imm0), ConstImm::Ieee64(imm1)) => {
+                if imm1 == 0.0 {
+                    return None;
+                }
+                Some(ConstImm::Ieee64(imm0 / imm1))
+            }
+            _ => unreachable!(),
+        },
+        _ => None,
+    }
+}
+
+/// Fold a numerical binary function.
+fn fold_numerical_binary(
+    dfg: &mut ir::DataFlowGraph,
+    inst: ir::Inst,
+    opcode: ir::Opcode,
+    args: [ir::Value; 2],
+) {
+    if let Some(const_imm) = evaluate_numerical_binary(dfg, opcode, args) {
+        use self::ConstImm::*;
+        match const_imm {
+            I64(imm) => {
+                let typevar = dfg.ctrl_typevar(inst);
+                dfg.replace(inst).iconst(typevar, imm);
+            }
+            Ieee32(imm) => {
+                dfg.replace(inst)
+                    .f32const(ir::immediates::Ieee32::with_bits(imm.to_bits()));
+            }
+            Ieee64(imm) => {
+                dfg.replace(inst)
+                    .f64const(ir::immediates::Ieee64::with_bits(imm.to_bits()));
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+fn fold_simple_branch(func: &mut ir::Function, inst: ir::Inst, opcode: ir::Opcode) {
+    let (cond, ebb, args) = {
+        let values = func.dfg.inst_args(inst);
+        let inst_data = &func.dfg[inst];
+        (
+            resolve_value_to_imm(&func.dfg, values[0]).unwrap(),
+            inst_data.branch_destination().unwrap(),
+            values[1..].to_vec(),
+        )
+    };
+
+    let truthiness = cond.evaluate_truthiness();
+    let branch_if_zero = match opcode {
+        ir::Opcode::Brz => true,
+        ir::Opcode::Brnz => false,
+        _ => unreachable!(),
+    };
+
+    if (branch_if_zero && !truthiness) || (!branch_if_zero && truthiness) {
+        func.dfg.replace(inst).jump(ebb, &args);
+        // remove the rest of the ebb to avoid verifier errors
+        while let Some(next_inst) = func.layout.next_inst(inst) {
+            func.layout.remove_inst(next_inst);
+        }
+    } else {
+        let mut pos = FuncCursor::new(func).at_inst(inst);
+        pos.remove_inst_and_step_back();
```

Comment:
To avoid skipping over one instruction, should this operator on the `FuncCursor` from `fold_constants`?